packages/data/src/components/use-select/index.js

```diff
@@ -0,0 +1,167 @@
+ * }
+ *
+ * // Rendered in the application:
+ * // <HammerPriceDisplay currency="USD" />
+ * ```
+ *
+ * In the above example, when `HammerPriceDisplay` is rendered into an
+ * application, the price will be retrieved from the store state using the
+ * `mapSelect` callback on `useSelect`. If the currency prop changes then
+ * any price in the state for that currency is retrieved. If the currency prop
+ * doesn't change and other props are passed in that do change, the price will
+ * not change because the dependency is just the currency.
+ *
+ * @return {Function}  A custom react hook.
+ */
+export default function useSelect( _mapSelect, deps ) {
+	const mapSelect = useCallback( _mapSelect, deps );
+	const registry = useRegistry();
+	const isAsync = useAsyncMode();
+	const queueContext = useMemo( () => ( { queue: true } ), [ registry ] );
+	const [ , forceRender ] = useReducer( ( s ) => s + 1, 0 );
+
+	const latestMapSelect = useRef();
+	const latestIsAsync = useRef( isAsync );
+	const latestMapOutput = useRef();
+	const latestMapOutputError = useRef();
+	const isMounted = useRef();
+
+	let mapOutput;
+
+	try {
+		if (
+			latestMapSelect.current !== mapSelect ||
+			latestMapOutputError.current
+		) {
+			mapOutput = mapSelect( registry.select, registry );
+		} else {
+			mapOutput = latestMapOutput.current;
+		}
+	} catch ( error ) {
+		let errorMessage = `An error occurred while running 'mapSelect': ${ error.message }`;
+
+		if ( latestMapOutputError.current ) {
+			errorMessage += `\nThe error may be correlated with this previous error:\n`;
+			errorMessage += `${ latestMapOutputError.current.stack }\n\n`;
+			errorMessage += 'Original stack trace:';
+
+			throw new Error( errorMessage );
+		}
+	}
+
+	useIsomorphicLayoutEffect( () => {
+		latestMapSelect.current = mapSelect;
+		if ( latestIsAsync.current !== isAsync ) {
+			latestIsAsync.current = isAsync;
+			renderQueue.flush( queueContext );
+		}
+		latestMapOutput.current = mapOutput;
+		latestMapOutputError.current = undefined;
+		isMounted.current = true;
+	} );
+
+	useIsomorphicLayoutEffect( () => {
+		const onStoreChange = () => {
+			if ( isMounted.current ) {
+				try {
+					const newMapOutput = latestMapSelect.current(
+						registry.select,
+						registry
+					);
+					if ( isShallowEqualObjects( latestMapOutput.current, newMapOutput ) ) {
+						return;
+					}
+					latestMapOutput.current = newMapOutput;
+				} catch ( error ) {
+					latestMapOutputError.current = error;
+				}
+				forceRender( {} );
+			}
+		};
+
+		// catch any possible state changes during mount before the subscription
+		// could be set.
+		if ( latestIsAsync.current ) {
+			renderQueue.add( queueContext, onStoreChange );
+		} else {
+			onStoreChange();
+		}
+
+		const unsubscribe = registry.subscribe( () => {
+			if ( latestIsAsync.current ) {
+				renderQueue.add( queueContext, onStoreChange );
+			} else {
+				onStoreChange();
+			}
+		} );
+
+		return () => {
+			isMounted.current = false;
```

Comment:
I don't know that it ends up being an issue, but this callback won't only be called for unmounting, it will also be called on a change in registry (and, if #19205 lands, in change in `isAsync`). I guess it would depend on the order of how `onStoreChange` or a subsequent call to the previous `useIsomorphicLayoutEffect` which sets `isMounted.current` back to `true`. My worry could be that `onStoreChange` wouldn't run the selector if it considers the component unmounted. I guess for the purpose of the selector running as unique to the registry (or `isAsync` in #19205), we don't want the invalidated change callback, and we can feel confident that either `renderQueue.add` or `onStoreChange` would be called for the next registry? In which case, it's more an issue of a confusing name.

Reply:
>I guess for the purpose of the selector running as unique to the registry (or isAsync in #19205), we don't want the invalidated change callback, and we can feel confident that either renderQueue.add or onStoreChange would be called for the next registry?

Exactly

>In which case, it's more an issue of a confusing name.

I am not sure what else to call it. `isCleaningUp`, `isUnsubscribing`?

Reply:
> > In which case, it's more an issue of a confusing name.
> 
> I am not sure what else to call it. `isCleaningUp`, `isUnsubscribing`?

Yeah, either of those would work pretty well I think (flipping the boolean value).

Reply:
#19286

Reply:
That boolean covers both scenarios^^